src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ReferralService.kt

```diff
@@ -204,6 +207,13 @@ class ReferralService(
       }
     }
 
+    update.serviceCategoryIds?.let {
+      val serviceCategories = serviceCategoryRepository.findByIdIn(it)
+      if (!referral.intervention.dynamicFrameworkContract.contractType.serviceCategories.containsAll(serviceCategories)) {
+        errors.add(FieldError(field = "serviceCategories", error = Code.INVALID_SERVICE_CATEGORY_FOR_CONTRACT))
+      }
+    }
+
```

Comment:
I don't think its necessary to fetch the serviceCategories since you can just to the comparison on the ids directly:

referral.intervention.dynamicFrameworkContract.contractType.serviceCategories.map{it.id}.containsAll(serviceCategoryIds)